**Problem:**

You are given the following dataset:
1. **Audible Data** : https://1drv.ms/u/s!AiqdXCxPTydhoog8ckLN-6Cw55fzIg?e=EWgZ5d

Your task is to:
- Find the problems with the datasets.
- Define the Data Quality Dimensions.
- Try to clean the datasets.

In [5]:
import pandas as pd
import numpy as np

In [202]:
audible=pd.read_excel('audible_uncleaned.xlsx')

In [7]:
audible

name                     author  \
0                      Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                                The Burning Maze      Writtenby:RickRiordan   
2                                    The Deep End       Writtenby:JeffKinney   
3                            Daughter of the Deep      Writtenby:RickRiordan   
4      The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   
...                                           ...                        ...   
87484                   Last Days of the Bus Club     Writtenby:ChrisStewart   
87485                                    The Alps    Writtenby:StephenO'Shea   
87486                        The Innocents Abroad        Writtenby:MarkTwain   
87487                       A Sentimental Journey   Writtenby:LaurenceSterne   
87488                                      Havana    Writtenby:MarkKurlansky   

                        narrator                time          releasedate  \
0          Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1       Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2          Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3      Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4      Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   
...                          ...                 ...                  ...   
87484    Narratedby:ChrisStewart   7 hrs and 34 mins  2017-09-03 00:00:00   
87485      Narratedby:RobertFass   10 hrs and 7 mins             21-02-17   
87486       Narratedby:FloGibson   19 hrs and 4 mins             30-12-16   
87487     Narratedby:AntonLesser    4 hrs and 8 mins             23-02-11   
87488     Narratedby:FleetCooper     6 hrs and 1 min  2017-07-03 00:00:00   

      language                          stars price  
0      English     5 out of 5 stars34 ratings   468  
1      English   4.5 out of 5 stars41 ratings   820  
2      English   4.5 out of 5 stars38 ratings   410  
3      English   4.5 out of 5 stars12 ratings   615  
4      English  4.5 out of 5 stars181 ratings   820  
...        ...                            ...   ...  
87484  English                  Not rated yet   596  
87485  English                  Not rated yet   820  
87486  English                  Not rated yet   938  
87487  English                  Not rated yet   680  
87488  English                  Not rated yet   569  

[87489 rows x 8 columns]

# 1. data summary

audible dataset

- data about 87489 audiobooks offered by Amazon as Audible service 

# 2. Column discription

1. `name`: name of the book
2. `author`: writer of the book
3. `narrator`: narrator of the book
4. `time`: total time of audiobook
5. `releasedate`: date when book was released
6. `language`: language of narration
7. `price` : amount of the book

In [8]:
audible.columns

Index(['name', 'author', 'narrator', 'time', 'releasedate', 'language',
       'stars', 'price'],
      dtype='object')

# 3. additional info

- there can be more than one author or narrator
- author can a person's name or a company's name
- foreword author are also present
- some narrator include middlename

Types of Assessment
There are 2 types of assessment styles

Manual - Looking through the data manually in google sheets
Programmatic - By using pandas functions such as info(), describe() or sample()

# 4. manual assessment

- check by column
- then by row

# issues with Data

1. Dirty Data
  - author - remove suffix writtenby `validity` [done]
  - narrator - remove suffix narratedby `validity` [done]
  - releasedate - year given both in yy and in yyyy formate [convert all to yyyy] `consistency` [done]
  - time - remove hrs and mins unit `validity` [done]
  - star - remove 'out of _ starts' and 'ratings' | 'rating' `validity` [done]
  - price - given in object - convert to float `accuracy` [done]
  - star - most (72417) movies are 'not rated yet' -> replace with nan `completeness` [revisit]
  - releasedate - some dates include time [but it's set to 00:00:00] `consistency` [done]
  - language - convert to a category `accuracy` [done]
  - language - some in lower case and some in capitalized format `consistency`
  - star - convert avg_rating to float and total_rating to int `accuracy` [done]
  - price - contains string 'free' instade of 0 `validity` [done]
  - total_ratings - remove ',' `validity` [done]


  - confusion in issues
  - [maybe an issue] name - contains '#' - could means number [can replace with no.]of just a hash tag [maybe part of original title - should not be changes] `validity`
  - [maybe an issue] name, author, narrator, releasedate,langauge, price - duplicated but time and rating for those audiobooks are different (96 rows) `consistency`
  - name, author, narrator - some data is corrupted `validity`
  - time - convert to correct hr:min format 
  - time datatype `accuracy`
  - author, narrator, name - can be converted to category `accuracy`
  - author - middle name sometimes are present, written in abbrivation or in fullform `consistency` 
  
2. Messy Data

  - star - star and total_ratings in same column [seperate] [done]
  - seperate time into hours and minutes [done]
  - [maybe an issuse]author - more than one author are present  
  - [maybe an issuse]narrator - more than one narrator are present
  

# 5. Automatic assessment

## full

### head

In [9]:
audible.duplicated().sum()

0

## 1. name

### head, tail & sample

- name corrupted

In [10]:
audible['name'].head()

0                    Geronimo Stilton #11 & #12
1                              The Burning Maze
2                                  The Deep End
3                          Daughter of the Deep
4    The Lightning Thief: Percy Jackson, Book 1
Name: name, dtype: object

In [11]:
audible['name'].tail()

87484    Last Days of the Bus Club
87485                     The Alps
87486         The Innocents Abroad
87487        A Sentimental Journey
87488                       Havana
Name: name, dtype: object

In [12]:
audible.sample(5)

name  \
68374           Lezioni sulla divina umanità   
7683               I'm Terrified of Bath Time   
19278  Living Biographies of Famous Novelists   
26155                        Unnatural Causes   
70654                 The Sunday Potluck Club   

                                        author                      narrator  \
68374              Writtenby:VladimirSolov'ëv     Narratedby:SilviaCecchini   
7683   Writtenby:SimonRich,TomToro-illustrator          Narratedby:SimonRich   
19278      Writtenby:HenryThomas,DanaLeeThomas         Narratedby:PatBottino   
26155              Writtenby:DrRichardShepherd  Narratedby:DrRichardShepherd   
70654                   Writtenby:MelissaStorm    Narratedby:StephanieCozart   

                     time          releasedate language  \
68374   7 hrs and 13 mins  2017-01-06 00:00:00  italian   
7683               4 mins  2022-05-04 00:00:00  English   
19278  10 hrs and 42 mins  2000-12-06 00:00:00  English   
26155  11 hrs and 46 mins             20-09-18  English   
70654   7 hrs and 43 mins             31-03-20  English   

                          stars price  
68374             Not rated yet   113  
7683              Not rated yet   500  
19278             Not rated yet   773  
26155  5 out of 5 stars1 rating   820  
70654             Not rated yet   703

In [13]:
audible['name'].count()-audible['name'].str.isalpha().sum()

80551

In [14]:
audible['name'].str.isalpha().replace(np.nan,False)

0        False
1        False
2        False
3        False
4        False
         ...  
87484    False
87485    False
87486    False
87487    False
87488     True
Name: name, Length: 87489, dtype: bool

In [15]:
import re

In [16]:
print(audible.iloc[15689]['name'])
print(audible.iloc[15689]['name'].isalpha())

第355回 新刊ラジオ第2部プレミアム　
False


In [17]:
audible.iloc[15687:]

name  \
15687  第352回 新刊ラジオ第2部プレミアãƒ...   
15688  第842回 新刊ラジオ第2部プレミアãƒ...   
15689  第355回 新刊ラジオ第2部プレミアãƒ...   
15690  第156回 新刊ラジオ第2部プレミアãƒ...   
15691  第65回 新刊ラジオ第2部プレミアム　   
...                                                  ...   
87484                          Last Days of the Bus Club   
87485                                           The Alps   
87486                               The Innocents Abroad   
87487                              A Sentimental Journey   
87488                                             Havana   

                                 author                           narrator  \
15687  Writtenby:矢島雅弘,石橋遊  Narratedby:矢島雅弘,石橋遊   
15688  Writtenby:矢島雅弘,石橋遊  Narratedby:矢島雅弘,石橋遊   
15689  Writtenby:矢島雅弘,石橋遊  Narratedby:矢島雅弘,石橋遊   
15690  Writtenby:矢島雅弘,石橋遊  Narratedby:矢島雅弘,石橋遊   
15691  Writtenby:矢島雅弘,石橋遊  Narratedby:矢島雅弘,石橋遊   
...                                 ...                                ...   
87484            Writtenby:ChrisStewart            Narratedby:ChrisStewart   
87485           Writtenby:StephenO'Shea              Narratedby:RobertFass   
87486               Writtenby:MarkTwain               Narratedby:FloGibson   
87487          Writtenby:LaurenceSterne             Narratedby:AntonLesser   
87488           Writtenby:MarkKurlansky             Narratedby:FleetCooper   

                    time          releasedate  language          stars price  
15687            29 mins             16-05-18  japanese  Not rated yet   139  
15688            32 mins             16-05-18  japanese  Not rated yet   139  
15689            16 mins             15-05-18  japanese  Not rated yet   139  
15690            25 mins             16-05-18  japanese  Not rated yet   139  
15691            18 mins             16-05-18  japanese  Not rated yet   139  
...                  ...                  ...       ...            ...   ...  
87484  7 hrs and 34 mins  2017-09-03 00:00:00   English  Not rated yet   596  
87485  10 hrs and 7 mins             21-02-17   English  Not rated yet   820  
87486  19 hrs and 4 mins             30-12-16   English  Not rated yet   938  
87487   4 hrs and 8 mins             23-02-11   English  Not rated yet   680  
87488    6 hrs and 1 min  2017-07-03 00:00:00   English  Not rated yet   569  

[71802 rows x 8 columns]

### info

In [18]:
audible['name'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: name
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


### isnull

In [19]:
audible['name'].isnull().sum()

0

In [20]:
audible[audible['name'].str.contains('#').replace(np.nan,False)]

name                     author  \
0        Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
15     Geronimo Stilton #13 and #14  Writtenby:GeronimoStilton   
27     Geronimo Stilton #20 and #21  Writtenby:GeronimoStilton   
30             Geronimo Stilton #17  Writtenby:GeronimoStilton   
36     Geronimo Stilton #15 and #16  Writtenby:GeronimoStilton   
...                             ...                        ...   
85017               #GoodGirlBadBoy    Writtenby:YeseniaVargas   
85045                  #HateThatBoy    Writtenby:YeseniaVargas   
85210            #TheRealCinderella    Writtenby:YeseniaVargas   
85822   Art Mages of Lure Series #1      Writtenby:JordanRivet   
85900               Eat, Pray, #FML   Writtenby:GabrielleStone   

                        narrator               time          releasedate  \
0          Narratedby:BillLobely  2 hrs and 20 mins  2008-04-08 00:00:00   
15         Narratedby:BillLobley  2 hrs and 25 mins  2008-08-02 00:00:00   
27         Narratedby:BillLobley  2 hrs and 32 mins  2010-01-10 00:00:00   
30         Narratedby:BillLobley  2 hrs and 28 mins             13-07-09   
36         Narratedby:BillLobley  2 hrs and 43 mins  2008-08-02 00:00:00   
...                          ...                ...                  ...   
85017     Narratedby:KylaGarcía  4 hrs and 26 mins             27-08-19   
85045     Narratedby:KylaGarcía  6 hrs and 12 mins  2019-06-08 00:00:00   
85210     Narratedby:KylaGarcía  6 hrs and 29 mins  2019-09-07 00:00:00   
85822     Narratedby:MarnyeYoung   9 hrs and 2 mins  2021-06-04 00:00:00   
85900  Narratedby:GabrielleStone  9 hrs and 39 mins  2020-11-02 00:00:00   

      language                         stars price  
0      English    5 out of 5 stars34 ratings   468  
15     English  4.5 out of 5 stars33 ratings   467  
27     English    5 out of 5 stars18 ratings   469  
30     English     5 out of 5 stars8 ratings   468  
36     English    4 out of 5 stars28 ratings   467  
...        ...                           ...   ...  
85017  English                 Not rated yet   469  
85045  English                 Not rated yet   586  
85210  English                 Not rated yet   586  
85822  English                 Not rated yet   586  
85900  English   2.5 out of 5 stars2 ratings   703  

[78 rows x 8 columns]

### duplicated

- not duplicate as same book is recited by many narrators
- check for duplicated for name,book and auther combined

In [21]:
audible['name'].duplicated().sum()

4722

In [22]:
audible[audible['name'].duplicated()]

name  \
18                      Merlin Mission Collection   
46     The Lightning Thief: Percy Jackson, Book 1   
54                      Merlin Mission Collection   
155                        Barbie - Superprincesa   
273                           Barbie - Dreamtopia   
...                                           ...   
87443       Travels with a Donkey in the Cevennes   
87456                                  Gettysburg   
87457                                        Solo   
87475                                  Wanderlust   
87486                        The Innocents Abroad   

                                           author  \
18                      Writtenby:MaryPopeOsborne   
46                          Writtenby:RickRiordan   
54                      Writtenby:MaryPopeOsborne   
155                              Writtenby:Mattel   
273    Writtenby:Mattel,MartaCisaMuñoz-traductor   
...                                           ...   
87443              Writtenby:RobertLouisStevenson   
87456                        Writtenby:JeffShaara   
87457                          Writtenby:PenHadow   
87475                    Writtenby:ElisabethEaves   
87486                         Writtenby:MarkTwain   

                             narrator                time  \
18         Narratedby:MaryPopeOsborne  10 hrs and 18 mins   
46             Narratedby:WalterLewis   3 hrs and 51 mins   
54         Narratedby:MaryPopeOsborne  13 hrs and 24 mins   
155    Narratedby:VanessaPérezJurado             29 mins   
273          Narratedby:MiriamMonlleo             21 mins   
...                               ...                 ...   
87443          Narratedby:DenisLawson   2 hrs and 51 mins   
87456        Narratedby:RobertsonDean    1 hr and 12 mins   
87457             Narratedby:PenHadow    3 hrs and 5 mins   
87475          Narratedby:ErinBennett   9 hrs and 51 mins   
87486            Narratedby:FloGibson   19 hrs and 4 mins   

               releasedate language                       stars price  
18     2017-02-05 00:00:00  English  5 out of 5 stars11 ratings  1256  
46                13-01-10  English   4 out of 5 stars4 ratings   615  
54     2017-02-05 00:00:00  English   5 out of 5 stars8 ratings  1256  
155    2022-07-03 00:00:00  spanish               Not rated yet    38  
273               31-01-22  catalan               Not rated yet    38  
...                    ...      ...                         ...   ...  
87443             13-05-08  English               Not rated yet   569  
87456             26-03-07  English               Not rated yet   200  
87457             14-02-05  English               Not rated yet   615  
87475             26-02-14  English               Not rated yet   668  
87486             30-12-16  English               Not rated yet   938  

[4722 rows x 8 columns]

In [23]:
audible.query('name=="Merlin Mission Collection"')
#not duplicate as same book is recited by many narrators

name                     author  \
8   Merlin Mission Collection  Writtenby:MaryPopeOsborne   
18  Merlin Mission Collection  Writtenby:MaryPopeOsborne   
54  Merlin Mission Collection  Writtenby:MaryPopeOsborne   

                      narrator                time          releasedate  \
8   Narratedby:MaryPopeOsborne  10 hrs and 56 mins  2017-02-05 00:00:00   
18  Narratedby:MaryPopeOsborne  10 hrs and 18 mins  2017-02-05 00:00:00   
54  Narratedby:MaryPopeOsborne  13 hrs and 24 mins  2017-02-05 00:00:00   

   language                       stars price  
8   English   5 out of 5 stars5 ratings  1256  
18  English  5 out of 5 stars11 ratings  1256  
54  English   5 out of 5 stars8 ratings  1256

### duplicated for name,author and narrator combined

In [24]:
audible[audible[['name','author','narrator']].duplicated()]

name                     author  \
18     Merlin Mission Collection  Writtenby:MaryPopeOsborne   
54     Merlin Mission Collection  Writtenby:MaryPopeOsborne   
1882          Egg Marks the Spot    Writtenby:AmyTimberlake   
2162                        Fish    Writtenby:LisaJ.Amstutz   
2244     The Snail and the Whale   Writtenby:JuliaDonaldson   
...                          ...                        ...   
86656                     Venice        Writtenby:JanMorris   
86725           Audiowalk Berlin     Writtenby:TaufigKhalil   
86746                  City Surf         Writtenby:CitySurf   
86747                  City Surf         Writtenby:CitySurf   
86749                  City Surf         Writtenby:CitySurf   

                           narrator                time          releasedate  \
18       Narratedby:MaryPopeOsborne  10 hrs and 18 mins  2017-02-05 00:00:00   
54       Narratedby:MaryPopeOsborne  13 hrs and 24 mins  2017-02-05 00:00:00   
1882      Narratedby:MichaelBoatman   3 hrs and 54 mins             14-09-21   
2162           Narratedby:anonymous             18 mins             23-02-22   
2244      Narratedby:ImeldaStaunton             22 mins             22-06-21   
...                             ...                 ...                  ...   
86656  Narratedby:SebastianComberti   5 hrs and 16 mins  2010-09-08 00:00:00   
86725       Narratedby:TaufigKhalil    1 hr and 12 mins  2005-01-06 00:00:00   
86746         Narratedby:uncredited             32 mins             23-10-07   
86747         Narratedby:uncredited             52 mins             23-10-07   
86749         Narratedby:uncredited             40 mins             23-10-07   

      language                       stars price  
18     English  5 out of 5 stars11 ratings  1256  
54     English   5 out of 5 stars8 ratings  1256  
1882   English               Not rated yet   398  
2162   English               Not rated yet    93  
2244   English               Not rated yet   251  
...        ...                         ...   ...  
86656  English               Not rated yet   668  
86725   german               Not rated yet   334  
86746  English               Not rated yet   234  
86747  English               Not rated yet   234  
86749  English               Not rated yet   234  

[404 rows x 8 columns]

In [25]:
audible[audible[['name','author','narrator','releasedate', 'language', 'price']].duplicated()]

name                      author  \
18                   Merlin Mission Collection   Writtenby:MaryPopeOsborne   
54                   Merlin Mission Collection   Writtenby:MaryPopeOsborne   
4321                   My First French Lessons     Writtenby:AlexaPolidoro   
4322                   My First French Lessons     Writtenby:AlexaPolidoro   
4337                   My First French Lessons     Writtenby:AlexaPolidoro   
...                                        ...                         ...   
81524  Gesänge und Rufe heimischer Vogelarten  Writtenby:KarlHeinzDingler   
86244        Japan, für dich ein Spaziergang!       Writtenby:RonjaSakata   
86746                                City Surf          Writtenby:CitySurf   
86747                                City Surf          Writtenby:CitySurf   
86749                                City Surf          Writtenby:CitySurf   

                         narrator                time          releasedate  \
18     Narratedby:MaryPopeOsborne  10 hrs and 18 mins  2017-02-05 00:00:00   
54     Narratedby:MaryPopeOsborne  13 hrs and 24 mins  2017-02-05 00:00:00   
4321     Narratedby:AlexaPolidoro             59 mins             28-11-06   
4322     Narratedby:AlexaPolidoro                1 hr             28-11-06   
4337     Narratedby:AlexaPolidoro             51 mins             28-11-06   
...                           ...                 ...                  ...   
81524             Narratedby:div.    1 hr and 15 mins  2011-12-10 00:00:00   
86244      Narratedby:RonjaSakata   3 hrs and 48 mins             20-10-19   
86746       Narratedby:uncredited             32 mins             23-10-07   
86747       Narratedby:uncredited             52 mins             23-10-07   
86749       Narratedby:uncredited             40 mins             23-10-07   

      language                       stars price  
18     English  5 out of 5 stars11 ratings  1256  
54     English   5 out of 5 stars8 ratings  1256  
4321   English               Not rated yet   190  
4322   English               Not rated yet   190  
4337   English               Not rated yet   190  
...        ...                         ...   ...  
81524   german               Not rated yet   200  
86244   german               Not rated yet   535  
86746  English               Not rated yet   234  
86747  English               Not rated yet   234  
86749  English               Not rated yet   234  

[96 rows x 8 columns]

### unique

In [26]:
len(audible['name'].unique())

82767

In [27]:
len(audible['name'])

87489

## author

### head, tail & sample

In [28]:
audible['author'].head()

0    Writtenby:GeronimoStilton
1        Writtenby:RickRiordan
2         Writtenby:JeffKinney
3        Writtenby:RickRiordan
4        Writtenby:RickRiordan
Name: author, dtype: object

In [29]:
audible['author'].tail()

87484      Writtenby:ChrisStewart
87485     Writtenby:StephenO'Shea
87486         Writtenby:MarkTwain
87487    Writtenby:LaurenceSterne
87488     Writtenby:MarkKurlansky
Name: author, dtype: object

In [30]:
audible['author'].sample(5)

75610         Writtenby:I.T.Lucas
18865      Writtenby:A.E.Hotchner
63740         Writtenby:Epictetus
16894         Writtenby:BillBrown
63467    Writtenby:TrevorAaronson
Name: author, dtype: object

### info

In [31]:
audible['author'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: author
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


### isnull

In [32]:
audible['author'].isnull().sum()

0

### duplicated

In [33]:
audible['author'].duplicated().sum()

39115

In [34]:
audible[audible['author'].duplicated()]
#same author can have many books -> not duplication

name                    author  \
3                            Daughter of the Deep     Writtenby:RickRiordan   
4      The Lightning Thief: Percy Jackson, Book 1     Writtenby:RickRiordan   
7                               The Dark Prophecy     Writtenby:RickRiordan   
9                             The Tyrant’s Tomb     Writtenby:RickRiordan   
10       The Titan's Curse: Percy Jackson, Book 3     Writtenby:RickRiordan   
...                                           ...                       ...   
87480                         Journey into Cyprus    Writtenby:ColinThubron   
87481                         The Hills of Adonis    Writtenby:ColinThubron   
87486                        The Innocents Abroad       Writtenby:MarkTwain   
87487                       A Sentimental Journey  Writtenby:LaurenceSterne   
87488                                      Havana   Writtenby:MarkKurlansky   

                        narrator                time          releasedate  \
3      Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4      Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   
7       Narratedby:RobbieDaymond  12 hrs and 32 mins  2017-02-05 00:00:00   
9       Narratedby:RobbieDaymond  13 hrs and 22 mins             24-09-19   
10     Narratedby:JesseBernstein   8 hrs and 48 mins             14-01-10   
...                          ...                 ...                  ...   
87480  Narratedby:JonathanKeeble  11 hrs and 51 mins  2018-01-01 00:00:00   
87481  Narratedby:JonathanKeeble   6 hrs and 52 mins  2017-01-12 00:00:00   
87486       Narratedby:FloGibson   19 hrs and 4 mins             30-12-16   
87487     Narratedby:AntonLesser    4 hrs and 8 mins             23-02-11   
87488     Narratedby:FleetCooper     6 hrs and 1 min  2017-07-03 00:00:00   

      language                          stars price  
3      English   4.5 out of 5 stars12 ratings   615  
4      English  4.5 out of 5 stars181 ratings   820  
7      English     5 out of 5 stars50 ratings   820  
9      English     5 out of 5 stars58 ratings   820  
10     English  4.5 out of 5 stars130 ratings   820  
...        ...                            ...   ...  
87480  English                  Not rated yet   785  
87481  English                  Not rated yet   679  
87486  English                  Not rated yet   938  
87487  English                  Not rated yet   680  
87488  English                  Not rated yet   569  

[39115 rows x 8 columns]

### unique
- can convert to a cateory

In [35]:
len(audible['author'].unique())

48374

## narrator

### head, tail & sample

In [36]:
audible['narrator'].head()

0        Narratedby:BillLobely
1     Narratedby:RobbieDaymond
2        Narratedby:DanRussell
3    Narratedby:SoneelaNankani
4    Narratedby:JesseBernstein
Name: narrator, dtype: object

In [37]:
audible['narrator'].tail()

87484    Narratedby:ChrisStewart
87485      Narratedby:RobertFass
87486       Narratedby:FloGibson
87487     Narratedby:AntonLesser
87488     Narratedby:FleetCooper
Name: narrator, dtype: object

In [38]:
audible['narrator'].sample(5)

43500    Narratedby:IggyToma,KirtGraves
14561           Narratedby:JulienAllouf
36694          Narratedby:PatrickLawlor
72495            Narratedby:OlgaDianova
9586            Narratedby:SoniaManzano
Name: narrator, dtype: object

### info

In [39]:
audible['narrator'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: narrator
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


### unique

In [40]:
len(audible['narrator'].unique())

29717

## time

### head, tail & sample

In [41]:
audible['time'].head()

0     2 hrs and 20 mins
1     13 hrs and 8 mins
2      2 hrs and 3 mins
3    11 hrs and 16 mins
4                10 hrs
Name: time, dtype: object

In [42]:
audible['time'].tail()

87484    7 hrs and 34 mins
87485    10 hrs and 7 mins
87486    19 hrs and 4 mins
87487     4 hrs and 8 mins
87488      6 hrs and 1 min
Name: time, dtype: object

In [43]:
audible['time'].sample(5)

39359    2 hrs and 34 mins
82319     6 hrs and 9 mins
70698    9 hrs and 17 mins
7569      2 hrs and 6 mins
31748    6 hrs and 10 mins
Name: time, dtype: object

### info

In [44]:
audible['time'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: time
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


### duplicated

In [45]:
audible['time'].duplicated().sum()

85205

In [46]:
audible[audible['time'].duplicated()]['time']

48         2 hrs and 1 min
67       5 hrs and 27 mins
70       2 hrs and 23 mins
76       5 hrs and 41 mins
84       9 hrs and 30 mins
               ...        
87484    7 hrs and 34 mins
87485    10 hrs and 7 mins
87486    19 hrs and 4 mins
87487     4 hrs and 8 mins
87488      6 hrs and 1 min
Name: time, Length: 85205, dtype: object

## releasedate

### head,tail & sample

In [47]:
audible['releasedate'].head()

0    2008-04-08 00:00:00
1    2018-01-05 00:00:00
2    2020-06-11 00:00:00
3    2021-05-10 00:00:00
4    2010-01-13 00:00:00
Name: releasedate, dtype: object

In [48]:
audible['releasedate'].tail()

87484    2017-09-03 00:00:00
87485               21-02-17
87486               30-12-16
87487               23-02-11
87488    2017-07-03 00:00:00
Name: releasedate, dtype: object

In [49]:
audible['releasedate'].sample(5)

49789    2008-05-12 00:00:00
4722     2022-04-01 00:00:00
82982    2021-01-12 00:00:00
18502               19-10-21
35569               22-03-22
Name: releasedate, dtype: object

### info

In [50]:
audible['releasedate'].info(5)

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: releasedate
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


### duplicated

In [51]:
audible['releasedate'].duplicated().sum()

82430

In [52]:
audible[audible['releasedate'].duplicated()]['releasedate']

8        2017-02-05 00:00:00
18       2017-02-05 00:00:00
19       2018-06-11 00:00:00
29                  25-11-14
36       2008-08-02 00:00:00
                ...         
87484    2017-09-03 00:00:00
87485               21-02-17
87486               30-12-16
87487               23-02-11
87488    2017-07-03 00:00:00
Name: releasedate, Length: 82430, dtype: object

## language


### unique

In [53]:
audible['language'].unique()

array(['English', 'Hindi', 'spanish', 'german', 'french', 'catalan',
       'swedish', 'italian', 'danish', 'finnish', 'dutch', 'hebrew',
       'russian', 'polish', 'galician', 'afrikaans', 'icelandic',
       'romanian', 'japanese', 'tamil', 'portuguese', 'urdu', 'hungarian',
       'czech', 'bulgarian', 'mandarin_chinese', 'basque', 'korean',
       'arabic', 'greek', 'turkish', 'ukrainian', 'slovene', 'norwegian',
       'telugu', 'lithuanian'], dtype=object)

### duplicated

In [54]:
audible['language'].duplicated().sum()

87453

### head, tail & sample

In [55]:
audible['language'].head()

0    English
1    English
2    English
3    English
4    English
Name: language, dtype: object

In [56]:
audible['language'].tail()

87484    English
87485    English
87486    English
87487    English
87488    English
Name: language, dtype: object

In [57]:
audible['language'].sample(5)

19771    English
59726    italian
24039     german
38995    English
26961    English
Name: language, dtype: object

In [58]:
audible['language'].value_counts()

English             61884
german               8295
spanish              3496
japanese             3167
italian              2694
french               2386
russian              1804
danish                935
portuguese            526
swedish               515
Hindi                 436
polish                224
finnish               197
dutch                 190
tamil                 161
catalan               153
mandarin_chinese       97
icelandic              52
romanian               50
hungarian              36
urdu                   34
afrikaans              28
czech                  23
turkish                20
greek                  18
arabic                 16
norwegian              16
galician               10
bulgarian               9
korean                  4
slovene                 4
hebrew                  2
basque                  2
telugu                  2
lithuanian              2
ukrainian               1
Name: language, dtype: int64

### info

In [59]:
audible['language'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: language
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


## stars

### head, tail & sample

In [60]:
audible['stars'].head()

0       5 out of 5 stars34 ratings
1     4.5 out of 5 stars41 ratings
2     4.5 out of 5 stars38 ratings
3     4.5 out of 5 stars12 ratings
4    4.5 out of 5 stars181 ratings
Name: stars, dtype: object

In [61]:
audible['stars'].tail()

87484    Not rated yet
87485    Not rated yet
87486    Not rated yet
87487    Not rated yet
87488    Not rated yet
Name: stars, dtype: object

In [62]:
audible['stars'].sample(5)

73614                  Not rated yet
82719                  Not rated yet
84328                  Not rated yet
87402                  Not rated yet
19717    4.5 out of 5 stars2 ratings
Name: stars, dtype: object

In [63]:
audible[audible['stars']=="Not rated yet"].count()

name           72417
author         72417
narrator       72417
time           72417
releasedate    72417
language       72417
stars          72417
price          72417
dtype: int64

### info

In [64]:
audible['stars'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: stars
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


## price

### head, tail & sample

In [65]:
audible['price'].head()

0    468
1    820
2    410
3    615
4    820
Name: price, dtype: object

In [66]:
audible['price'].tail()

87484    596
87485    820
87486    938
87487    680
87488    569
Name: price, dtype: object

In [67]:
audible['price'].sample(5)

18768    267
79330    139
81870    352
27072    586
48017    930
Name: price, dtype: object

### info

- convert object to float

In [68]:
audible['price'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: price
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


### describe

- apply after converting it to int

# create copy of original dataset

- making changes in copy dataset and not in the original one

In [69]:
audible_copy=audible.copy()

In [70]:
audible_copy

name                     author  \
0                      Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                                The Burning Maze      Writtenby:RickRiordan   
2                                    The Deep End       Writtenby:JeffKinney   
3                            Daughter of the Deep      Writtenby:RickRiordan   
4      The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   
...                                           ...                        ...   
87484                   Last Days of the Bus Club     Writtenby:ChrisStewart   
87485                                    The Alps    Writtenby:StephenO'Shea   
87486                        The Innocents Abroad        Writtenby:MarkTwain   
87487                       A Sentimental Journey   Writtenby:LaurenceSterne   
87488                                      Havana    Writtenby:MarkKurlansky   

                        narrator                time          releasedate  \
0          Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1       Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2          Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3      Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4      Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   
...                          ...                 ...                  ...   
87484    Narratedby:ChrisStewart   7 hrs and 34 mins  2017-09-03 00:00:00   
87485      Narratedby:RobertFass   10 hrs and 7 mins             21-02-17   
87486       Narratedby:FloGibson   19 hrs and 4 mins             30-12-16   
87487     Narratedby:AntonLesser    4 hrs and 8 mins             23-02-11   
87488     Narratedby:FleetCooper     6 hrs and 1 min  2017-07-03 00:00:00   

      language                          stars price  
0      English     5 out of 5 stars34 ratings   468  
1      English   4.5 out of 5 stars41 ratings   820  
2      English   4.5 out of 5 stars38 ratings   410  
3      English   4.5 out of 5 stars12 ratings   615  
4      English  4.5 out of 5 stars181 ratings   820  
...        ...                            ...   ...  
87484  English                  Not rated yet   596  
87485  English                  Not rated yet   820  
87486  English                  Not rated yet   938  
87487  English                  Not rated yet   680  
87488  English                  Not rated yet   569  

[87489 rows x 8 columns]

# Data cleaning

## 1. Completeness




### 1) star
most (72417) movies are 'not rated yet' -> replace with nan - completeness

In [71]:
#1. star - most (72417) movies are completeness

#check 
audible_copy[audible_copy['stars']=='Not rated yet']['stars'].count()

#define - replace  'not rated yet' with nan
#after creating seperate cols for star and total_rating -> coming back again 

72417

In [72]:
#code
audible_copy['stars']=audible_copy['stars'].replace('Not rated yet',np.nan)

In [73]:
#test
audible_copy[audible_copy['stars']=='Not rated yet']['stars'].count()

0

### 2) total_ratings

- replacing NaN with 0

In [161]:
audible_copy['total_ratings'].isna().sum()

72417

In [180]:
audible_copy['total_ratings']=audible_copy['total_ratings'].fillna(0)

In [181]:
audible_copy['total_ratings'].isna().sum()

0

## 2. messy data








In [74]:
#1. star and total_ratings in same column [seperate]

#check
audible_copy['stars'].head(5)
#define: seprate total_rating(ratings) and average_rating(stars)

0       5 out of 5 stars34 ratings
1     4.5 out of 5 stars41 ratings
2     4.5 out of 5 stars38 ratings
3     4.5 out of 5 stars12 ratings
4    4.5 out of 5 stars181 ratings
Name: stars, dtype: object

In [75]:
#code
audible_copy['star']=audible_copy['stars'].str.split('stars').str.get(0)
audible_copy['total_ratings']=audible_copy['stars'].str.split('stars').str.get(1)

In [76]:
#test
audible_copy.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language                          stars price           star total_ratings  
0  English     5 out of 5 stars34 ratings   468    5 out of 5     34 ratings  
1  English   4.5 out of 5 stars41 ratings   820  4.5 out of 5     41 ratings  
2  English   4.5 out of 5 stars38 ratings   410  4.5 out of 5     38 ratings  
3  English   4.5 out of 5 stars12 ratings   615  4.5 out of 5     12 ratings  
4  English  4.5 out of 5 stars181 ratings   820  4.5 out of 5    181 ratings

In [77]:
audible_copy.drop(columns='stars',inplace=True)

In [78]:
#test
audible_copy.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language price           star total_ratings  
0  English   468    5 out of 5     34 ratings  
1  English   820  4.5 out of 5     41 ratings  
2  English   410  4.5 out of 5     38 ratings  
3  English   615  4.5 out of 5     12 ratings  
4  English   820  4.5 out of 5    181 ratings

### 2

create new columns for date and time

In [79]:
def convert_to_time(time):
  temp=time.split('and') 
  hour='0'
  min='0'
  if(len(temp)==1):
    if('hrs' in temp[0]):
      hour=temp[0]
    else:
      min=temp[0]
  else:
    hour=temp[0]
    min=temp[1]
 
  return hour,min

In [80]:
total_time=audible_copy['time'].apply(lambda x: convert_to_time(x))

In [81]:
audible_copy['hours']=total_time.str.get(0)
audible_copy['minutes']=total_time.str.get(1)

In [82]:
audible_copy.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time          releasedate  \
0      Narratedby:BillLobely   2 hrs and 20 mins  2008-04-08 00:00:00   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins  2018-01-05 00:00:00   
2      Narratedby:DanRussell    2 hrs and 3 mins  2020-06-11 00:00:00   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins  2021-05-10 00:00:00   
4  Narratedby:JesseBernstein              10 hrs  2010-01-13 00:00:00   

  language price           star total_ratings    hours   minutes  
0  English   468    5 out of 5     34 ratings   2 hrs    20 mins  
1  English   820  4.5 out of 5     41 ratings  13 hrs     8 mins  
2  English   410  4.5 out of 5     38 ratings   2 hrs     3 mins  
3  English   615  4.5 out of 5     12 ratings  11 hrs    16 mins  
4  English   820  4.5 out of 5    181 ratings   10 hrs         0

## 3. validity issues

### 1) author - remove suffix  

In [83]:
#author - remove suffix writtenby validity

#check
audible_copy['author'].head()

0    Writtenby:GeronimoStilton
1        Writtenby:RickRiordan
2         Writtenby:JeffKinney
3        Writtenby:RickRiordan
4        Writtenby:RickRiordan
Name: author, dtype: object

In [84]:
print(audible_copy['author'].str.contains('Writtenby:').sum(),audible.shape)
#issue for all rows

87489 (87489, 8)


In [85]:
#define: replace Writtenby: with ''

#code
audible_copy['author']=audible_copy['author'].str.replace('Writtenby:','')

In [86]:
#test
audible_copy['author'].str.contains('Writtenby:').sum()

0

### 2) narrator - remove suffix   

In [87]:
#check
audible_copy['narrator'].head()

0        Narratedby:BillLobely
1     Narratedby:RobbieDaymond
2        Narratedby:DanRussell
3    Narratedby:SoneelaNankani
4    Narratedby:JesseBernstein
Name: narrator, dtype: object

In [88]:
print(audible_copy['narrator'].str.contains('Narratedby:').sum(),audible.shape)
#issue for all rows

87489 (87489, 8)


In [89]:
#define: replace Narratedby: with ''

#code
audible_copy['narrator']=audible_copy['narrator'].str.replace('Narratedby:','')

In [90]:
#test
audible_copy['narrator'].str.contains('Narratedby:').sum()

0

### 3) time - remove hrs and mins unit 
 

In [91]:
#check
audible_copy[['hours','minutes']].head()

hours   minutes
0   2 hrs    20 mins
1  13 hrs     8 mins
2   2 hrs     3 mins
3  11 hrs    16 mins
4   10 hrs         0

In [92]:
#define - replace hrs and mins with ''
audible_copy['minutes']=audible_copy['minutes'].str.replace('mins','').str.replace('min','').str.strip()

In [93]:
audible_copy['hours']=audible_copy['hours'].str.replace('hrs','').str.replace('hs','').str.strip()

In [94]:
#test
audible_copy[['hours','minutes']]

hours minutes
0         2      20
1        13       8
2         2       3
3        11      16
4        10       0
...     ...     ...
87484     7      34
87485    10       7
87486    19       4
87487     4       8
87488     6       1

[87489 rows x 2 columns]

### 4. star - remove suffinx

star - remove 'out of _ starts' and 'ratings' `validity`

In [95]:
# check
audible_copy[['star',	'total_ratings']].head()

star total_ratings
0    5 out of 5     34 ratings
1  4.5 out of 5     41 ratings
2  4.5 out of 5     38 ratings
3  4.5 out of 5     12 ratings
4  4.5 out of 5    181 ratings

In [154]:
#define replace it with ''
audible_copy['star']=audible_copy['star'].str.replace('out of 5','').str.strip()
audible_copy['total_ratings']=audible_copy['total_ratings'].str.replace('ratings','').str.replace('rating','').str.strip()

In [98]:
#test
audible_copy[['star',	'total_ratings']].head()

star total_ratings
0    5            34
1  4.5            41
2  4.5            38
3  4.5            12
4  4.5           181

In [155]:
audible_copy['total_ratings'].str.contains('rating').sum()

0

### 5) price - 'free' instade of 0 


In [117]:
#check
audible_copy[audible_copy['price'].str.contains('Free',na=False)]['price'].count()

338

In [128]:
#define: replace 'Free' with 0

#code
audible_copy['price']=audible_copy['price'].replace('Free','0')

In [129]:
#test
audible_copy['price'].str.contains('Free',na=False).sum()

0

6) total_ratings - remove ','

In [166]:
#check
audible_copy['total_ratings'].str.contains(',').sum()

39

In [171]:
audible_copy[audible_copy['total_ratings'].str.contains(',',na=False)]['total_ratings'].head()

7551     1,423
7555     1,006
19432    1,216
22158    1,823
22174    1,636
Name: total_ratings, dtype: object

In [173]:
#define - replace ',' with ''
#code
audible_copy['total_ratings']=audible_copy['total_ratings'].str.replace(',','')

In [174]:
#test
audible_copy['total_ratings'].str.contains(',').sum()

0

## Accuracy issues

### 1) price in float

price - given in object - convert to float accuracy

In [130]:
#check
audible_copy['price'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: price
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


In [131]:
audible_copy['price'].sample(5)

48537    727
23805    820
85554    820
32710     33
13950    233
Name: price, dtype: object

In [132]:
audible_copy['price'].iloc[983]

'0'

In [135]:
#define: convert object type to float
#test
audible_copy['price']=audible_copy['price'].astype(float)

### 2) language - convert to category

In [138]:
#check
audible_copy['language'].value_counts()

English             61884
german               8295
spanish              3496
japanese             3167
italian              2694
french               2386
russian              1804
danish                935
portuguese            526
swedish               515
Hindi                 436
polish                224
finnish               197
dutch                 190
tamil                 161
catalan               153
mandarin_chinese       97
icelandic              52
romanian               50
hungarian              36
urdu                   34
afrikaans              28
czech                  23
turkish                20
greek                  18
arabic                 16
norwegian              16
galician               10
bulgarian               9
korean                  4
slovene                 4
hebrew                  2
basque                  2
telugu                  2
lithuanian              2
ukrainian               1
Name: language, dtype: int64

In [141]:
audible_copy['language'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: language
Non-Null Count  Dtype 
--------------  ----- 
87489 non-null  object
dtypes: object(1)
memory usage: 683.6+ KB


In [142]:
#define: converting object to category using astype
audible_copy['language']=audible_copy['language'].astype('category')

In [143]:
#test
audible_copy['language'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: language
Non-Null Count  Dtype   
--------------  -----   
87489 non-null  category
dtypes: category(1)
memory usage: 86.9 KB


### 3)star - convert datatype

avg_rating to float and total_rating to int

In [144]:
#check
audible_copy[['star','total_ratings']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   star           15072 non-null  object
 1   total_ratings  15072 non-null  object
dtypes: object(2)
memory usage: 1.3+ MB


In [149]:
audible_copy[['star','total_ratings']].head(5)

star total_ratings
0    5            34
1  4.5            41
2  4.5            38
3  4.5            12
4  4.5           181

In [183]:
#define 
#convert star to float
#convert total_ratings to int

#code
audible_copy['star']=audible_copy['star'].astype('float')
audible_copy['total_ratings']=audible_copy['total_ratings'].astype('int')

In [182]:
audible_copy['total_ratings'].isna().sum()

0

## Consistency

### 1) releasedate - contains both yy and yyyy

 
[convert all to yyyy] 


### 2) releasedate - remove time part

In [190]:
#check
audible_copy['releasedate'].head()

0    2008-04-08 00:00:00
1    2018-01-05 00:00:00
2    2020-06-11 00:00:00
3    2021-05-10 00:00:00
4    2010-01-13 00:00:00
Name: releasedate, dtype: object

In [204]:
#define converting all to yyyy format
audible_copy['releasedate']=pd.to_datetime(audible_copy['releasedate'])
audible_copy['releasedate']=audible_copy['releasedate'].dt.strftime('%d-%m-%Y')

In [207]:
#test
audible_copy['releasedate'].sample(5)

30951    05-01-2018
39901    30-03-2020
83974    30-12-2007
15323    03-02-2018
75884    19-01-2021
Name: releasedate, dtype: object

In [211]:
audible_copy['releasedate'].to_csv('releasedate.csv')

In [208]:
#check
audible_copy['releasedate'].head()

0    08-04-2008
1    05-01-2018
2    11-06-2020
3    10-05-2021
4    13-01-2010
Name: releasedate, dtype: object

3) language - capitalized

In [213]:
#check 
audible_copy['language'].value_counts()

English             61884
german               8295
spanish              3496
japanese             3167
italian              2694
french               2386
russian              1804
danish                935
portuguese            526
swedish               515
Hindi                 436
polish                224
finnish               197
dutch                 190
tamil                 161
catalan               153
mandarin_chinese       97
icelandic              52
romanian               50
hungarian              36
urdu                   34
afrikaans              28
czech                  23
turkish                20
greek                  18
arabic                 16
norwegian              16
galician               10
bulgarian               9
slovene                 4
korean                  4
lithuanian              2
hebrew                  2
telugu                  2
basque                  2
ukrainian               1
Name: language, dtype: int64

In [216]:
#define capitalize all

#code
audible_copy['language']=audible_copy['language'].str.capitalize()

In [217]:
#test
audible_copy['language'].value_counts()

English             61884
German               8295
Spanish              3496
Japanese             3167
Italian              2694
French               2386
Russian              1804
Danish                935
Portuguese            526
Swedish               515
Hindi                 436
Polish                224
Finnish               197
Dutch                 190
Tamil                 161
Catalan               153
Mandarin_chinese       97
Icelandic              52
Romanian               50
Hungarian              36
Urdu                   34
Afrikaans              28
Czech                  23
Turkish                20
Greek                  18
Arabic                 16
Norwegian              16
Galician               10
Bulgarian               9
Korean                  4
Slovene                 4
Hebrew                  2
Basque                  2
Telugu                  2
Lithuanian              2
Ukrainian               1
Name: language, dtype: int64

In [220]:
audible_copy['language']=audible_copy['language'].astype('category')

In [221]:
audible_copy['language'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 87489 entries, 0 to 87488
Series name: language
Non-Null Count  Dtype   
--------------  -----   
87489 non-null  category
dtypes: category(1)
memory usage: 86.9 KB
